chirp/drivers/baofeng_digital.py

```diff
@@ -0,0 +1,492 @@
+            raise
+        except Exception as e:
+            raise errors.RadioError(
+                f'Failed to communicate with the radio: {e}')
+        finally:
+            self._exit_programming()
+
+    def process_mmap(self):
+        mem_fmt = self.MEM_FORMAT.format(chan_count=self._proto.chan_count)
+        self._memobj = bitwise.parse(mem_fmt, self._mmap)
+
+    def get_raw_memory(self, number):
+        return repr(self._memobj.memory[number])
+
+    def _parse_tone(self, value):
+        if value in (0x0000, 0xFFFF):
+            return ('', None, 'N')
+
+        # Parse lower 14 bit as BCD
+        bcd_value = (
+            1000 * (value >> 12 & 0x3) +
+            100 * (value >> 8 & 0xF) +
+            10 * (value >> 4 & 0xF) +
+            (value & 0xF)
+        )
+
+        # DTCS if bit 15 is set
+        if value & 0x8000:
+            return ('DTCS', bcd_value, value & 0x4000 and 'R' or 'N')
+        else:
+            # Parse as CTCSS frequency * 10
+            return ('Tone', bcd_value / 10.0, 'N')
+
+    def get_memory(self, number):
+        raw = self._memobj.memory[number - 1]
+
+        mem = chirp_common.Memory()
+        mem.number = number
+        mem.empty = raw.rx_freq.get_raw() in (
+            b'\x00\x00\x00\x00', b'\xFF\xFF\xFF\xFF'
+        )
+
+        bcl = hopping = False
+        encryption_key = 0
+        if not mem.empty:
+            rx_freq = int(raw.rx_freq) * 10
+            tx_freq = int(raw.tx_freq) * 10
+
+            mem.freq = rx_freq
+            offset = tx_freq - rx_freq
+            if offset == 0:
+                mem.duplex = ''
+                mem.offset = 0
+            elif abs(offset) < 10000000:
+                mem.duplex = '-' if offset < 0 else '+'
+                mem.offset = abs(offset)
+            else:
+                mem.duplex = 'split'
+                mem.offset = tx_freq
+
+            if raw.digital:
+                mem.mode = raw.narrow and 'DN' or 'DIG'
+                mem.tmode = ''
+                mem.dtcs_polarity = 'NN'
+                encryption_key = raw.encryption_key
+            else:
+                mem.mode = raw.narrow and 'NFM' or 'FM'
+                hopping = raw.hopping
+
+                tx_tmode, tx_tvalue, tx_tpol = self._parse_tone(raw.tx_tone)
+                if tx_tmode == 'Tone':
+                    mem.rtone = tx_tvalue
+                elif tx_tmode == 'DTCS':
+                    mem.dtcs = tx_tvalue
+
+                rx_tmode, rx_tvalue, rx_tpol = self._parse_tone(raw.rx_tone)
+                if rx_tmode == 'Tone':
+                    mem.ctone = rx_tvalue
+                elif rx_tmode == 'DTCS':
+                    mem.rx_dtcs = rx_tvalue
+
+                if tx_tmode == 'Tone' and rx_tmode == '':
+                    mem.tmode = 'Tone'
+                elif (
+                    tx_tmode and
+                    tx_tmode == rx_tmode and
+                    tx_tvalue == rx_tvalue
+                ):
```

Comment:
Please don't waste the two lines here. I know some people like black formatting, but I hate it :)